labs/Language/IntroTypeHackery.md

```diff
@@ -0,0 +1,257 @@
+`x` and `y` are both odd; then we'll compute either `x - y` or `y - x`
+and the symbolic simulator will think we have no knowledge of the bits
+of either difference. The subsequent `>> 1` will introduce a single
+`False` bit to our computation, but overall we will have decreased the
+certainty of the bits in our arguments as we recurse.
+
+The next trick is to explicitly remove the unnecessary bits from the
+arguments as we recurse. If we represent halving by trimming a bit
+from the argument, the symbolic simulator reckons that one or both of
+the arguments eventually have type `[0]` which serves as a base
+case. What we have done is change from reducing values to eventually
+reach a base case, to reducing the types until the base case is
+reached. After that it's all details, but there are several.
+
+
+Implementing Binary gcd in Cryptol
+----------------------------------
+
+We'll need a few supporting functions:
+
++ A test for even is needed for various sized values. It is
+  implemented in terms of `odd` below.
+
+```cryptol
+/**
+ * Indicates whether the argument is even or odd.
+ */
+
+even, odd : {n} (fin n) => [n] -> Bit
+even = ~odd
+odd x = last ([False] # x) // "[False] #" affords type [0]
+```
+
++ A halving function that also reduces the bit width by 1 if
+  possible. This is employed in the recursive call to make the size of
+  one or both arguments smaller. It turns out that `take` with the
+  right type restrictions does the trick.
+
+```cryptol
+/**
+ * Halves by discarding the last bit if possible.
+ */
+
+halve : {n} (fin n) => [n] -> [max n 1 - 1]
+halve = take
+```
+
++ Because the result of a recursive call with be 0 or 1 bit smaller
+  than the desired result type, `zext` is applied to all the recursive
+  results. In step 2, we also need to double the recursive result
+  which is easily achieved by `<< 1`. N.b., in the case where neither
+  argument is `0` and both are even, when we recurse the result will
+  necessarily be one bit skinnier, `zext` will add back a most
+  significant `False` bit and `<< 1` can not cause an overflow.
+
++ As a convenience, we define an absolute difference function whose
+  arguments may be of different bit widths.
+
+```
+/**
+ * Computes absolute difference with heterotypic arguments.
+ */
+
+absDiff : {m, n} (fin m, fin n) => [m] -> [n] -> [max m n]
+absDiff x y = if x' > y' then x' - y' else y' - x'
+  where
+    x', y' : [max m n]
+    x' = zext x
+    y' = zext y
+```
+
++ Since only one of the arguments may be halved at each recursive
+  step, the widths of the two arguments must be independent, giving us
+  the more general type signature:
+
+```cryptol
+gcd : {m, n} (fin m, fin n) => [m] -> [n] -> [max m n]
+```
+
+  We need the result type to be `[max m n]` since if either argument
+  is zero, the result is the value of the other argument.
+
+Without further adieu, the code:
+
+```cryptol
+gcd x y = if x == 0 \/ y == 0 then zext x ^ zext y        // line 1
```

Comment:
Since we now have a nice symbolically terminating function, will you write a property that demonstrates it?

Reply:
A monomorphic one.